pageserver/src/walredo.rs

```diff
@@ -49,6 +50,9 @@ use postgres_ffi::nonrelfile_utils::mx_offset_to_member_offset;
 use postgres_ffi::nonrelfile_utils::transaction_id_set_status;
 use postgres_ffi::pg_constants;
 
+/// Maximum number of WAL redo processes to launch for a single tenant.
+const MAX_PROCESSES: usize = 4;
+
```

Comment:
I wonder if instead of a fixed number of WAL redo processes we just need a global pool? this way we can dedicate more resources to the timelines in need.

Reply:
WAL redo processes cannot be shared across tenants for security reasons. An attacker could generate malicious WAL to hijack the WAL redo process, and if the same process was reused for another tenant, the hijacked process could see the other tenant's data.

We could have a global maximum limit on the number of processes though. So that if one tenant needs more WAL processess, we would kill a process of a less-active tenant to make room for it.

Reply:
Frankly speaking I do not think that 10% improvement at one workload can justify the increased complexity and especially increased number of processes which can in turn lead to overall degrade of performance.  With which number of clients this 10% improvement is reached?
And what about speed with just one client? In may experiments having two wal redo processes with round robin schedule cause quite noticeable decrease of performance.

Reply:
> Frankly speaking I do not think that 10% improvement at one workload can justify the increased complexity and especially increased number of processes which can in turn lead to overall degrade of performance. With which number of clients this 10% improvement is reached?

This is with 4 clients. I would expect more benefit with more clients, but haven't tested.

> And what about speed with just one client? In may experiments having two wal redo processes with round robin schedule cause quite noticeable decrease of performance.

Interesting. This uses the redo processes in LIFO order, not round robin, so that should not be an issue. But I guess we should add a single-client test to our test suite too.